# Enter Name and ID

## Name: Aleksandra Olshanova

## ID: 728945613

# Online Retail Exercise

In this exercise you will be given some data about the transactions of a UK-based and registered non-store online retail.  An invoice number starting with the letter 'c' indicates a cancelled order.

Some data may not make sense.

Follow the directions and try your best to answer the questions and complete the tasks.

**Import pandas and read in the Online Retail csv file and set it to a DataFrame called retail**  Use the parameter *encoding= 'unicode_escape'* for read_csv() method.

In [38]:
import pandas as pd 

In [39]:
retail = pd.read_csv("Online Retail.csv", encoding= 'unicode_escape')
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       01/12/2010 08:26       2.55     17850.0  United Kingdom  
1       01/12/2010 08:26       3.39     17850.0  United Kingdom  
2       01/12/2010 08:26       2.75     17850.0  United Kingdom  
3       01/12/2010 08:26       3.39     17850.0  United Kingdom  
4       01/12/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  09/12/2011 12:50       0.85     12680.0          France  
541905  09/12/2011 12:50       2.10     12680.0          France  
541906  09/12/2011 12:50       4.15     12680.0          France  
541907  09/12/2011 12:50       4.15     12680.0          France  
541908  09/12/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

**1.**  Check the head of the DataFrame. 

In [40]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  01/12/2010 08:26       2.55     17850.0  United Kingdom  
1  01/12/2010 08:26       3.39     17850.0  United Kingdom  
2  01/12/2010 08:26       2.75     17850.0  United Kingdom  
3  01/12/2010 08:26       3.39     17850.0  United Kingdom  
4  01/12/2010 08:26       3.39     17850.0  United Kingdom

**2.** Use the .info() method to find out how many entries there are.

In [41]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


**3.** Convert InvoiceDate column type to the right type.

In [42]:
retail['InvoiceDate'] = retail['InvoiceDate'].astype('datetime64')
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-01-12 08:26:00       2.55     17850.0  United Kingdom  
1      2010-01-12 08:26:00       3.39     17850.0  United Kingdom  
2      2010-01-12 08:26:00       2.75     17850.0  United Kingdom  
3      2010-01-12 08:26:00       3.39     17850.0  United Kingdom  
4      2010-01-12 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-09-12 12:50:00       0.85     12680.0          France  
541905 2011-09-12 12:50:00       2.10     12680.0          France  
541906 2011-09-12 12:50:00       4.15     12680.0          France  
541907 2011-09-12 12:50:00       4.15     12680.0          France  
541908 2011-09-12 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

**4.** Notice that the CustomerID column has only 406829 entries while other columns have much more.  Use the .isnull() and .sum() to find the total number of missing values in all of the dataframe.

In [43]:
null_bool = retail.isnull().sum()
#null_bool
total_sum = null_bool['Description'].sum() + null_bool['CustomerID'].sum()
total_sum

136534

**5.** What are the highest and lowest prices per unit for the diffrent items?

In [44]:
UnitPrice_min = retail['UnitPrice'].min()
UnitPrice_max = retail['UnitPrice'].max()
print('the highest unit price is {} and the lowest is {}'.format(UnitPrice_max, UnitPrice_min))

the highest unit price is 38970.0 and the lowest is -11062.06


**6.** Find the rows with UnitPrice that lower than zero.

In [45]:
retail.loc[retail.UnitPrice < 0]
#retail[retail['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299983   A563186         B  Adjust bad debt         1 2011-12-08 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-12-08 14:52:00   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

**7.** Find the actual price of those items and fix the prices. Show the rows after the correction.

In [46]:
items = retail[(retail['Description'] == 'Adjust bad debt') & (retail['UnitPrice'] > 0)]
items

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299982   A563185         B  Adjust bad debt         1 2011-12-08 14:50:00   

        UnitPrice  CustomerID         Country  
299982   11062.06         NaN  United Kingdom

In [47]:
A = items.index
real_price = items.loc[A]['UnitPrice'].values[0]
real_price

11062.06

In [48]:
wrong_UnitPrice_index = list(retail.loc[retail.UnitPrice < 0].index)
wrong_UnitPrice_index 

[299983, 299984]

In [49]:
for x in wrong_UnitPrice_index: 
 retail.loc[x, ['UnitPrice']] = real_price
retail.loc[wrong_UnitPrice_index[0]:wrong_UnitPrice_index[-1]]

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299983   A563186         B  Adjust bad debt         1 2011-12-08 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-12-08 14:52:00   

        UnitPrice  CustomerID         Country  
299983   11062.06         NaN  United Kingdom  
299984   11062.06         NaN  United Kingdom

Show  Dataframe that includes all of the rows with UnitPrice that between 200-300.

In [50]:
retail.loc[(retail.UnitPrice > 200) & (retail.UnitPrice < 300)]
#retail[(retail['UnitPrice'] > 200) & (retail['UnitPrice']< 300)]

InvoiceNo StockCode                  Description  Quantity  \
4989      536835     22655  VINTAGE RED KITCHEN CABINET         1   
5450      536862       DOT               DOTCOM POSTAGE         1   
14498    C537597         D                     Discount        -1   
15879     537641       DOT               DOTCOM POSTAGE         1   
19392    C537857         D                     Discount        -1   
...          ...       ...                          ...       ...   
438979   C574344      POST                      POSTAGE        -1   
519508    580148       DOT               DOTCOM POSTAGE         1   
526015    580645         M                       Manual         1   
535003    581179      POST                      POSTAGE         1   
541541   C581499         M                       Manual        -1   

               InvoiceDate  UnitPrice  CustomerID         Country  
4989   2010-02-12 18:06:00     295.00     13145.0  United Kingdom  
5450   2010-03-12 11:13:00     254.43         NaN  United Kingdom  
14498  2010-07-12 12:34:00     281.00     15498.0  United Kingdom  
15879  2010-07-12 15:32:00     229.74         NaN  United Kingdom  
19392  2010-08-12 16:00:00     267.12     17340.0  United Kingdom  
...                    ...        ...         ...             ...  
438979 2011-04-11 10:18:00     262.73     12415.0       Australia  
519508 2011-02-12 10:08:00     274.25         NaN  United Kingdom  
526015 2011-05-12 13:11:00     219.50     17857.0  United Kingdom  
535003 2011-07-12 15:43:00     240.00     12471.0         Germany  
541541 2011-09-12 10:28:00     224.69     15498.0  United Kingdom  

[207 rows x 8 columns]

**9.** Find the 3 moments with higest amount of items purchased. !!!!!!

In [51]:
A = retail['InvoiceDate'].value_counts(ascending=False)
A.head(3)

2011-10-31 14:41:00    1114
2011-08-12 09:28:00     749
2011-09-12 10:03:00     731
Name: InvoiceDate, dtype: int64

**10.** How many times people bought more than 10 of a single item from the United Kingdom?

In [52]:
United_Kingdom_data = retail[(retail['Country'] == 'United Kingdom') & (retail['Quantity'] > 10)]
United_Kingdom_data.shape[0]

109870

**11. We would now like to determine the percentage of every the total invoice For each CustomerID.  We will do this in two ways: using merging and using transform:**

**a.** First, create a new column with the total price for each item purchased (based on quantity and unit price)

In [53]:
retail['Total Price'] = retail['Quantity']*retail['UnitPrice']
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  Total Price  
0      2010-01-12 08:26:00       2.55     17850.0  United Kingdom        15.30  
1      2010-01-12 08:26:00       3.39     17850.0  United Kingdom        20.34  
2      2010-01-12 08:26:00       2.75     17850.0  United Kingdom        22.00  
3      2010-01-12 08:26:00       3.39     17850.0  United Kingdom        20.34  
4      2010-01-12 08:26:00       3.39     17850.0  United Kingdom        20.34  
...                    ...        ...         ...             ...          ...  
541904 2011-09-12 12:50:00       0.85     12680.0          France        10.20  
541905 2011-09-12 12:50:00       2.10     12680.0          France        12.60  
541906 2011-09-12 12:50:00       4.15     12680.0          France        16.60  
541907 2011-09-12 12:50:00       4.15     12680.0          France        16.60  
541908 2011-09-12 12:50:00       4.95     12680.0          France        14.85  

[541909 rows x 9 columns]

**b.** Find the total price of each invoice using grouby.

In [54]:
order_total = retail.groupby('InvoiceNo')['Total Price'].sum()
order_total


InvoiceNo
536365        139.12
536366         22.20
536367        278.73
536368         70.05
536369         17.85
             ...    
C581484   -168469.60
C581490       -32.53
C581499      -224.69
C581568       -54.75
C581569        -7.50
Name: Total Price, Length: 25900, dtype: float64

**c.** Change the name of the Series to "Order Total" and then reset the index of the new data (so you will get a DataFrame). 

In [55]:
new_name = order_total.rename("Order Total")
#new_name
order_total_data = pd.DataFrame(new_name.reset_index())
order_total_data

InvoiceNo  Order Total
0        536365       139.12
1        536366        22.20
2        536367       278.73
3        536368        70.05
4        536369        17.85
...         ...          ...
25895   C581484   -168469.60
25896   C581490       -32.53
25897   C581499      -224.69
25898   C581568       -54.75
25899   C581569        -7.50

[25900 rows x 2 columns]

**d.** Use Merge function to combine the new data with the original data to a new DataFrame called "new_retail". <br>Add a new column with the percentage of the order named Percentage of Order.

In [56]:
new_retail = retail.merge(order_total_data)
new_retail['Perentage of Order'] = new_retail['Total Price'] / new_retail['Order Total'] 
new_retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Total Price  \
0 2010-01-12 08:26:00       2.55     17850.0  United Kingdom        15.30   
1 2010-01-12 08:26:00       3.39     17850.0  United Kingdom        20.34   
2 2010-01-12 08:26:00       2.75     17850.0  United Kingdom        22.00   
3 2010-01-12 08:26:00       3.39     17850.0  United Kingdom        20.34   
4 2010-01-12 08:26:00       3.39     17850.0  United Kingdom        20.34   

   Order Total  Perentage of Order  
0       139.12            0.109977  
1       139.12            0.146205  
2       139.12            0.158137  
3       139.12            0.146205  
4       139.12            0.146205

**The second approach is to use groupby along with transform.  This avoids the issue of reordering the index to obtain the new dataframe.  We retain the same number of items as in the original data set.**<br>
**e.** Use groupby and transform function to make a new column - Order Total.

In [57]:
retail["Order Total"] = retail.groupby('InvoiceNo')['Total Price'].transform('sum')
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  \
0      2010-01-12 08:26:00       2.55     17850.0  United Kingdom   
1      2010-01-12 08:26:00       3.39     17850.0  United Kingdom   
2      2010-01-12 08:26:00       2.75     17850.0  United Kingdom   
3      2010-01-12 08:26:00       3.39     17850.0  United Kingdom   
4      2010-01-12 08:26:00       3.39     17850.0  United Kingdom   
...                    ...        ...         ...             ...   
541904 2011-09-12 12:50:00       0.85     12680.0          France   
541905 2011-09-12 12:50:00       2.10     12680.0          France   
541906 2011-09-12 12:50:00       4.15     12680.0          France   
541907 2011-09-12 12:50:00       4.15     12680.0          France   
541908 2011-09-12 12:50:00       4.95     12680.0          France   

        Total Price  Order Total  
0             15.30       139.12  
1             20.34       139.12  
2             22.00       139.12  
3             20.34       139.12  
4             20.34       139.12  
...             ...          ...  
541904        10.20       249.45  
541905        12.60       249.45  
541906        16.60       249.45  
541907        16.60       249.45  
541908        14.85       249.45  

[541909 rows x 10 columns]

**f.** Add a column to with the percentage of the order.

In [58]:
retail['Perentage of Order'] = retail['Total Price'] / retail['Order Total'] 
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  \
0      2010-01-12 08:26:00       2.55     17850.0  United Kingdom   
1      2010-01-12 08:26:00       3.39     17850.0  United Kingdom   
2      2010-01-12 08:26:00       2.75     17850.0  United Kingdom   
3      2010-01-12 08:26:00       3.39     17850.0  United Kingdom   
4      2010-01-12 08:26:00       3.39     17850.0  United Kingdom   
...                    ...        ...         ...             ...   
541904 2011-09-12 12:50:00       0.85     12680.0          France   
541905 2011-09-12 12:50:00       2.10     12680.0          France   
541906 2011-09-12 12:50:00       4.15     12680.0          France   
541907 2011-09-12 12:50:00       4.15     12680.0          France   
541908 2011-09-12 12:50:00       4.95     12680.0          France   

        Total Price  Order Total  Perentage of Order  
0             15.30       139.12            0.109977  
1             20.34       139.12            0.146205  
2             22.00       139.12            0.158137  
3             20.34       139.12            0.146205  
4             20.34       139.12            0.146205  
...             ...          ...                 ...  
541904        10.20       249.45            0.040890  
541905        12.60       249.45            0.050511  
541906        16.60       249.45            0.066546  
541907        16.60       249.45            0.066546  
541908        14.85       249.45            0.059531  

[541909 rows x 11 columns]

# Good luck!